### ISSUE TITLE
AnyEvent::CondVar exception is thrown when Facebook::Graph is used with Plack::Middleware

### ISSUE THREAD

**Issue by omani** [rizen/Facebook-Graph#37]
{
   "exception" : "AnyEvent::CondVar: recursive blocking wait attempted at /home/foo/perl5/perlbrew/perls/perl-5.16.0/lib/site_perl/5.16.0/Facebook/Graph/AccessToken.pm line 46.\n",
   "error" : "AnyEvent::CondVar: recursive blocking wait attempted at /home/foo/perl5/perlbrew/perls/perl-5.16.0/lib/site_perl/5.16.0/Facebook/Graph/AccessToken.pm line 46.\n"
}

If Perl Dancer is used with Plack::Middleware and is fired up with plackup, and Dancer::Plugin::Facebook is used (which uses Facebook::Graph), the above exception is thrown.

sub in Facebook/Graph/AccessToken.pm is:
  sub request {
      my ($self) = @_; 
      return Facebook::Graph::AccessToken::Response->new(
          response => Facebook::Graph::Request->new->get($self->uri_as_string)->recv->response
      );  
  }

line 46 is: Facebook::Graph::Request->new->get($self->uri_as_string)->recv->response

**Comment by davehodg** [comment]
Two years later, I'm getting this with Caztalyst and Facebook::Graph. Did you get any solution?

**Comment by jussikinnula** [comment]
I've comprehensively tested the way Facebook::Graph does requests at the moment (e.g. with AnyEvent and AnyEvent::HTTP and more specifically AnyEvent::HTTP::LWP::UserAgent wrapper), and found that there's a socket leak which makes constant use of FB logins and access_token requests fail. The error message I've got error message "Could not fetch access token: No such device or address at /app/local/lib/perl5/Facebook/Graph/AccessToken/Response.pm line 24." happening randomly on FB logins. I've used Catalyst middleware "Catalyst::Authentication::Credential::Facebook::OAuth2" for the logins.

I'm making a pull request which has a fix, which is already verified to work on couple of websites. Basically it's reverting back to LWP::UserAgent, and removing all AnyAgent related dependencies from the Facebook::Graph.
